### DISCUSSION TITLE
ch2/ch2-06

### DISCUSSION THREAD

**Discussion by giscus** [gopl-zh/gopl-zh.github.com#195] [General]
# ch2/ch2-06

<The Go Programming Language>中文版

https://gopl-zh.github.io/ch2/ch2-06.html

**Comment by zb81** [comment]
练习 2.3
```go
package main

import (
	"fmt"
	"time"
)

// 0-255 每个数值包含的 1 的位数（八位数值）
var pc = func() (pc [256]byte) {
	for i := range pc {
		pc[i] = pc[i>>1] + byte(i&1)
	}
	return
}()

func PopCount(x uint64) int {
	return int(pc[byte(x>>(0*8))] + // 最低 8 位
		pc[byte(x>>(1*8))] +
		pc[byte(x>>(2*8))] +
		pc[byte(x>>(3*8))] +
		pc[byte(x>>(4*8))] +
		pc[byte(x>>(5*8))] +
		pc[byte(x>>(6*8))] +
		pc[byte(x>>(7*8))]) // 最高 8 位
}

func main() {
	arr := make([]int, 10000000)
	for i := range arr {
		arr[i] = i + 1
	}
	start := time.Now()
	for _, v := range arr {
		PopCount(uint64(v))
	}
	fmt.Printf("pc cost: %d ms\n", time.Since(start).Milliseconds()) // 1ms

	start = time.Now()
	for _, v := range arr {
		PopCount2(uint64(v))
	}
	fmt.Printf("loop cost: %d ms\n", time.Since(start).Milliseconds()) // 48ms

}

func PopCount2(x uint64) int {
	count := 0
	for x != 0 {
		if x&1 == 1 {
			count++
		}
		x >>= 1
	}
	return count
}
```

**Reply by chenchenchenmo** [reply]
go真难写，没有python好写

**Comment by SuperZ3** [comment]
```go
package main

import (
	"fmt"
)

func main() {
	fmt.Println(PopCount(9999))
	fmt.Println(PopCountEach1(9999))
	fmt.Println(PopCountEach8(9999))
	fmt.Println(PopCount0(9999))
}

var pc [256]byte

func init() {
    for i := range pc {
        pc[i] = pc[i/2] + byte(i&1)
    }
}

// PopCount returns the population count (number of set bits) of x.
func PopCount(x uint64) int {
    return int(pc[byte(x>>(0*8))] +
        pc[byte(x>>(1*8))] +
        pc[byte(x>>(2*8))] +
        pc[byte(x>>(3*8))] +
        pc[byte(x>>(4*8))] +
        pc[byte(x>>(5*8))] +
        pc[byte(x>>(6*8))] +
        pc[byte(x>>(7*8))])
}

func PopCountEach1(x uint64) int {
	var res int
	for x > 0 {
		if x&1 > 0 {
			res++
		}
		x = x>>1
	}
	return res
}

func PopCountEach8(x uint64) int {
    var res int
	for step := 7; step >= 0; step-- {
		res += int(pc[byte(x>>(8 * step))])
	}
	return res
}

func PopCount0(x uint64) int {
	var res int
	for x > 0 {
		x &= x-1
		res++
	}
	return res
}
```

**Comment by Fi-tang** [comment]
onebitcount.go 
'''go
package onebitcount

var onebitMap [256]byte

func init(){
	for i, _ := range onebitMap{
		onebitMap[i] = onebitMap[i] + byte(i & 1)
	}
}

func OnebitFoundamental(x uint64) int {
	return int(
		onebitMap[byte(x >> ( 0 * 8))] + 
		onebitMap[byte(x >> ( 1 * 8))] + 
		onebitMap[byte(x >> ( 2 * 8))] + 
		onebitMap[byte(x >> ( 3 * 8))] + 
		onebitMap[byte(x >> ( 4 * 8))] + 
		onebitMap[byte(x >> ( 5 * 8))] + 
		onebitMap[byte(x >> ( 6 * 8))] +
		onebitMap[byte(x >> ( 7 * 8))],  
	)
}

func Onebit23(x uint64) int {
	var num int 
	for i := 0; i < 8; i++ {
		num += int(onebitMap[byte(x >> (i * 8))])
	}
	return num
}

func OnebitShift24(x uint64) int {
	var num int 
	for x != 0 {
		num += int(x & 1)
		x >>= 1
	}
	return num 
}

// x & (x - 1) 能够将 x 的最后一个非 零 bit 位清零
func OnebitClear25(x uint64) int {
	var count int 
	for x != 0 {
		x = x & (x - 1)
		count++
	}
	return count
}```
测试程序
```go
/*
练习 2.5:
表达式 x & (x - 1) 用于将 x 的最低一个非零的 bit 位清零。
使用这个算法重写 PopCount 函数，然后比较性能

结果: 统一来说
testFoundamental: 1.378473526 s
testLoop23: 18.059520553 s
testShift24: 24.736205763 s
testClear25: 18.931325779 s

总结来说，查表最快，其次将查表写成循环的速度约等于 x & (x - 1), 最慢的是每次移动最后一位 x >>= 1 的做法
*/
package main 

import (
	"fmt"
	"onebitcount"
	"time"
)

func testFoundamental(){
	start := time.Now()
	var i uint64
	for i = 0; i < 1024 * 1024 * 1024; i++ {
		onebitcount.OnebitFoundamental(i)
	}
	fmt.Printf("testFoundamental: %vs\n", time.Since(start).Seconds())
}

func test23(){
	start := time.Now()
	var i uint64
	for i = 0; i < 1024 * 1024 * 1024; i++ {
		onebitcount.Onebit23(i)
	}
	fmt.Printf("testLoop23: %vs\n", time.Since(start).Seconds())
}

func test24(){
	start := time.Now()
	var i uint64
	for i = 0; i < 1024 * 1024 * 1024; i++ {
		onebitcount.OnebitShift24(i)
	}
	fmt.Printf("testShift24: %vs\n", time.Since(start).Seconds())
}

func test25(){
	start := time.Now()
	var i uint64
	for i = 0; i < 1024 * 1024 * 1024; i++ {
		onebitcount.OnebitClear25(i)
	}
	fmt.Printf("testClear25: %vs\n", time.Since(start).Seconds())
}

func main(){
	testFoundamental()
	test23()
	test24()
	test25()
}
```
文件目录结构 onebitcount   / onebitcount.go 
                                                test 文件夹     test.go 
使用的编译命令行
在 onebitcount 文件夹里输入
```Bash
go mod init onebitcount
```
在 test 文件夹里输入
```Bash
go build test.go
```

**Comment by newarabbit** [comment]
2.2
// Package main
package main

import (
    "fmt"
    "os"
    "strconv"
)

type Meter float64
type Kilometer float64


func (m Meter) String() string      { return fmt.Sprintf("%gM", m) }
func (km Kilometer) String() string { return fmt.Sprintf("%gKM", km) }

// MToKm converts a Meter temperature to Kilometer.
func MToKm(m Meter) Kilometer { return Kilometer(m * 0.001) }

func KmToM(km Kilometer) Meter { return Meter(km * 1000) }

func main() {
    for _, arg := range os.Args[1:] {
       t, err := strconv.ParseFloat(arg, 64)
       if err != nil {
          fmt.Fprintf(os.Stderr, "cf: %v\n", err)
          os.Exit(1)
       }
       m := Meter(t)
       km := Kilometer(t)
       fmt.Printf("%s=%s,%s=%s\n\n", m, MToKm(m), km, KmToM(km))
    }
}

**Comment by lwc18** [comment]
package main

import (
	"fmt"
	"time"
)

// pc[i] is the population count of i.
var pc [256]byte

func init() {
	for i := range pc {
		pc[i] = pc[i/2] + byte(i&1)
	}
}

// PopCount1 returns the population count (number of set bits) of x.
func PopCount1(x uint64) int {
	return int(pc[byte(x>>(0*8))] +
		pc[byte(x>>(1*8))] +
		pc[byte(x>>(2*8))] +
		pc[byte(x>>(3*8))] +
		pc[byte(x>>(4*8))] +
		pc[byte(x>>(5*8))] +
		pc[byte(x>>(6*8))] +
		pc[byte(x>>(7*8))])
}

// PopCount2 returns the population count (number of set bits) of x.
func PopCount2(x uint64) int {
	var count int
	for x > 0 {
		x = x & (x - 1)
		count++
	}
	return count
}

func main() {
	const num = 1000000000
	x := uint64(1234567890)
	fmt.Println(PopCount1(x))
	fmt.Println(PopCount2(x))
	start := time.Now()
	for i := 0; i < num; i++ {
		PopCount1(x)
	}
	elapsed := time.Since(start)
	fmt.Printf("PopCount1 took %s\n", elapsed)

	start = time.Now()
	for i := 0; i < num; i++ {
		PopCount2(x)
	}
	elapsed = time.Since(start)
	fmt.Printf("PopCount2 took %s\n", elapsed)
}

**Comment by allenjlma** [comment]
这段翻译不太准确。
原文是：
> In Go, as in any other programming language, one builds large programs from a small set of basic constructs. Variables store values. Simple expressions are combined into larger ones with operations like addition and subtraction. Basic types are collected into aggregates like arrays and structs. Expressions are used in statements whose execution order is determined by control-ﬂow statements like if and for. Statements are grouped into functions for isolation and reuse. Functions are gathered into source ﬁles and packages

应该翻译成：
如很多其他编程语言一样，Go语言中的大程序都是从一小组基础组件构建而来：变量存储值；简单表达式通过加法和减法等操作组合成更大的表达式；基本类型被聚合成数组和结构体；表达式被语句（statements）所使用，这些语句（statements）的执行顺序由if和for等控制流语句来决定；为了隔离和复用，语句又被组织成函数；函数又更进一步被组织成源文件和包。

**Reply by allenjlma** [reply]
简单表达式->复杂表达式->语句->函数->包; 基础数据类型->聚合体 层层递进，组成大程序

**Comment by zhaoyanggh** [comment]
func PopCountLoop(x uint64) int {
	res := 0
	for i := 0; i < 8; i++ {
		res += int(Pc[byte(x>>(i*8))])
	}
	return res
}

func PopCountMove(x uint64) int {
	res := 0
	for x > 0 {
		if x%2 == 1 {
			res++
			x--
		}
		x /= 2
	}
	return res
}

func PopCountClean(x uint64) int {
	res := 0
	for x > 0 {
		x = x & (x - 1)
		res++
	}
	return res
}

**Reply by siqi47-a** [reply]
打卡

**Comment by Apahno** [comment]
GPT时代打卡

**Comment by skyfisher** [comment]
package main

import (
	"crypto/sha256"
	"fmt"
)

func count(c1 [32]byte, c2 [32]byte) int {
	cnt := 0
	for i := 0; i < len(c1); i++ {
		cnt += popcount(c1[i], c2[i])
	}

	return cnt
}

// 逐位进行比较
func popcount(x, y uint8) int {
	c := 0
	for i := 0; i < 8; i++ {
		fmt.Printf("%d,%d\n", x%2, y%2)
		if x%2 != y%2 {
			c += 1
		}
		x = x >> 1
		y = y >> 1
	}
	return c
}

func main() {
	c1 := sha256.Sum256([]byte("x"))
	c2 := sha256.Sum256([]byte("X"))
	fmt.Printf("%x\n%x\n%t\n%T\n", c1, c2, c1 == c2, c1)
	fmt.Printf("%d\n", count(c1, c2))
	// Output:
	// 2d711642b726b04401627ca9fbac32f5c8530fb1903cc4db02258717921a4881
	// 4b68ab3847feda7d6c62c1fbcbeebfa35eab7351ed5e78f4ddadea5df64b8015
	// false
	// [32]uint8
}

**Comment by leolin49** [comment]
```go
package main

import "fmt"
import "time"

var pc [256]byte = func() (p [256]byte) {
	for i := range p {
		p[i] = p[i>>1] + byte(i&1)
	}
	return
}()

func PopCount1(x uint64) int {
    return int(pc[byte(x>>(0*8))] +
        pc[byte(x>>(1*8))] +
        pc[byte(x>>(2*8))] +
        pc[byte(x>>(3*8))] +
        pc[byte(x>>(4*8))] +
        pc[byte(x>>(5*8))] +
        pc[byte(x>>(6*8))] +
        pc[byte(x>>(7*8))])
}

func PopCount2(x uint64) (cnt int) {
	for i := 0; i < 8; i++ {
		cnt += int(pc[byte(x>>(8*i))])
	}
	return
}

func PopCount3(x uint64) (cnt int) {
	for x > 0 {
		if x&1 > 0 {
			cnt++
		}
		x >>= 1
	}
	return
}

func PopCount4(x uint64) (cnt int) {
	for x > 0 {
		cnt++
		x &= x-1
	}
	return
}

func test1() {
	start := time.Now()
	var i uint64
	for i = 0; i < 1024 * 1024 * 1024; i++ {
		PopCount1(i)
	}
	fmt.Printf("test1: %vs\n", time.Since(start).Seconds())
}

func test2() {
	start := time.Now()
	var i uint64
	for i = 0; i < 1024 * 1024 * 1024; i++ {
		PopCount2(i)
	}
	fmt.Printf("test2: %vs\n", time.Since(start).Seconds())
}

func test3() {
	start := time.Now()
	var i uint64
	for i = 0; i < 1024 * 1024 * 1024; i++ {
		PopCount3(i)
	}
	fmt.Printf("test3: %vs\n", time.Since(start).Seconds())
}

func test4(){
	start := time.Now()
	var i uint64
	for i = 0; i < 1024 * 1024 * 1024; i++ {
		PopCount4(i)
	}
	fmt.Printf("test4: %vs\n", time.Since(start).Seconds())
}

func main(){
	test1()
	test2()
	test3()
	test4()
}

```

**Comment by Chen-05** [comment]
<pre>
pc.go
package popcount

// 查表方式-匿名函数版本
var pc = func() (pc [256]byte) {
	for i := range pc {
		pc[i] = pc[i>>1] + byte(i&1)
	}
	return
}()

// 查表方式-init 版本
//var pc [256]byte
//
//func init() {
//	for i := range pc {
//		pc[i] = pc[i>>1] + byte(i&1)
//	}
//}

func PopCountTable(x uint64) int {
	return int(pc[byte(x>>(0*8))] +
		pc[byte(x>>(1*8))] +
		pc[byte(x>>(2*8))] +
		pc[byte(x>>(3*8))] +
		pc[byte(x>>(4*8))] +
		pc[byte(x>>(5*8))] +
		pc[byte(x>>(6*8))] +
		pc[byte(x>>(7*8))])
}

func PopCountLoop(x uint64) int {
	count := 0
	for i := 0; i < 8; i++ {
		count += int(pc[byte(x>>(i*8))])
	}
	return count
}

// PopCountTable for 循环版本
func PopCountShift(x uint64) int {
	count := 0
	for x != 0 {
		count += int(x & 1)
		x = x >> 1
	}
	return count
}

// PopCountTable Brian Kernighan算法
func PopCountKernighan(x uint64) int {
	count := 0
	for x != 0 {
		x &= x - 1
		count++
	}
	return count
}

pc_test.go
package popcount

import "testing"

var x uint64 = 0x123456789ABCDEF

func BenchmarkPopCountTable(b *testing.B) {
	for i := 0; i < b.N; i++ {
		PopCountTable(x)
	}
}

func BenchmarkPopCountLoop(b *testing.B) {
	for i := 0; i < b.N; i++ {
		PopCountLoop(x)
	}
}

func BenchmarkPopCountShift(b *testing.B) {
	for i := 0; i < b.N; i++ {
		PopCountShift(x)
	}
}

func BenchmarkPopCountKernighan(b *testing.B) {
	for i := 0; i < b.N; i++ {
		PopCountKernighan(x)
	}
}

BenchmarkPopCountTable
BenchmarkPopCountTable-12               1000000000               0.2599 ns/op
BenchmarkPopCountLoop
BenchmarkPopCountLoop-12                253351470                4.769 ns/op
BenchmarkPopCountShift
BenchmarkPopCountShift-12               36166473                31.39 ns/op
BenchmarkPopCountKernighan
BenchmarkPopCountKernighan-12           69168250                17.90 ns/op
PASS

</pre>

**Comment by Fragmentmemory** [comment]
package popcount

// pc[i] is the population count of i.
var pc [256]byte = func() (pc [256]byte) {
	for i := range pc {
		pc[i] = pc[i/2] + byte(i&1)
	}
	return
}()

// func init() {
// 	for i := range pc {
// 		pc[i] = pc[i/2] + byte(i&1)
// 	}
// }

// PopCount returns the population count (number of set bits) of x.
func PopCount(x uint64) int {
	return int(pc[byte(x>>(0*8))] +
		pc[byte(x>>(1*8))] +
		pc[byte(x>>(2*8))] +
		pc[byte(x>>(3*8))] +
		pc[byte(x>>(4*8))] +
		pc[byte(x>>(5*8))] +
		pc[byte(x>>(6*8))] +
		pc[byte(x>>(7*8))])
}

func PopCount4Loop(x uint64) int {
	var count int
	for x > 0 {
		count += int(x & 1)
		x >>= 1
	}
	return count
}

func PopCount4Prime(x uint64) int {
	var count int
	for x > 0 {
		x &= x - 1
		count++
	}
	return count
}

// 测试程序
package main

import (
	"flag"
	"fmt"
	"time"

	"GoProject/learn/ch2/popcount"
)

func testTime(num uint64, op func(uint64) int, iterations int) {
	start := time.Now()
	var result int
	for i := 0; i < iterations; i++ {
		result = op(num) // 执行 N 次，但只保留最后一次结果
	}
	total := time.Since(start)
	avgNs := total.Nanoseconds() / int64(iterations)

	fmt.Printf("结果:%d ", result)
	fmt.Printf("执行 %d 次总耗时：%v\n", iterations, total)
	fmt.Printf("平均单次耗时：%d ns (%.6f ms)\n", avgNs, float64(avgNs)/1e6)
}

func main() {
	var num uint64

	flag.Uint64Var(&num, "num", 0, "计算1位的个数")

	flag.Parse()

	const iters = 100000 // 根据需要调整，保证总耗时 > 1ms
	testTime(num, popcount.PopCount, iters)
	testTime(num, popcount.PopCount4Loop, iters)
	testTime(num, popcount.PopCount4Prime, iters)
}
